SignalServiceKit/src/Storage/FullTextSearchFinder.swift

```diff
@@ -0,0 +1,141 @@
+        guard let ext: YapDatabaseFullTextSearchTransaction = ext(transaction: transaction) else {
+            assertionFailure("ext was unexpectedly nil")
+            return
+        }
+
+        let normalized = FullTextSearchFinder.normalize(text: searchText)
+
+        // We want a forgiving query for phone numbers
+        // TODO a stricter "whole word" query for body text?
+        let prefixQuery = "*\(normalized)*"
+
+        // (snippet: String, collection: String, key: String, object: Any, stop: UnsafeMutablePointer<ObjCBool>)
+        ext.enumerateKeysAndObjects(matching: prefixQuery, with: nil) { (snippet: String, _: String, _: String, object: Any, _: UnsafeMutablePointer<ObjCBool>) in
+            block(object, snippet)
+        }
+    }
+
+    private func ext(transaction: YapDatabaseReadTransaction) -> YapDatabaseFullTextSearchTransaction? {
+        return transaction.ext(FullTextSearchFinder.dbExtensionName) as? YapDatabaseFullTextSearchTransaction
+    }
+
+    // Mark: Index Building
+
+    private class var contactsManager: ContactsManagerProtocol {
+        return TextSecureKitEnv.shared().contactsManager
+    }
+
+    private class func normalize(text: String) -> String {
+        var normalized: String = text.trimmingCharacters(in: .whitespacesAndNewlines)
+
+        // Remove any phone number formatting from the search terms
+        let nonformattingScalars = normalized.unicodeScalars.lazy.filter {
+            !CharacterSet.punctuationCharacters.contains($0)
+        }
+
+        normalized = String(String.UnicodeScalarView(nonformattingScalars))
+
+        return normalized
+    }
+
+    private static let groupThreadIndexer: SearchIndexer<TSGroupThread> = SearchIndexer { (groupThread: TSGroupThread) in
+        let groupName = groupThread.groupModel.groupName ?? ""
+
+        let memberStrings = groupThread.groupModel.groupMemberIds.map { recipientId in
+            recipientIndexer.index(recipientId)
+        }.joined(separator: " ")
+
+        let searchableContent = "\(groupName) \(memberStrings)"
+
+        return normalize(text: searchableContent)
+    }
+
+    private static let contactThreadIndexer: SearchIndexer<TSContactThread> = SearchIndexer { (contactThread: TSContactThread) in
+        let recipientId =  contactThread.contactIdentifier()
+        let searchableContent = recipientIndexer.index(recipientId)
+
+        return normalize(text: searchableContent)
+    }
+
+    private static let recipientIndexer: SearchIndexer<String> = SearchIndexer { (recipientId: String) in
+        let displayName = contactsManager.displayName(forPhoneIdentifier: recipientId)
+        let searchableContent =  "\(recipientId) \(displayName)"
+
+        return normalize(text: searchableContent)
+    }
+
+    private static let messageIndexer: SearchIndexer<TSMessage> = SearchIndexer { (message: TSMessage) in
+        let searchableContent =  message.body ?? ""
```

Comment:
Should we add a TODO to include:

* text from oversize text attachments.
* text from quoted replies.
* etc.?

Reply:
I'm not compelled to index replies by the thing that they're quoting. It wouldn't be difficult from the search index standpoint, we'd just have the indexer include the text.

But showing the user what's matched, and disambiguating the quote from what was quoted seems difficult and dubiously useful.

Searching oversize text attachments seems useful and without the UI pain points, so I'll add a card, but they're not very commonly used, but I'd suggest it's lower priority work.